src/tools/rbd_mirror/image_sync/SnapshotCopyRequest.cc

```diff
@@ -167,7 +179,19 @@ void SnapshotCopyRequest<I>::handle_snap_unprotect(int r) {
   CephContext *cct = m_local_image_ctx->cct;
   ldout(cct, 20) << ": r=" << r << dendl;
 
-  if (r < 0) {
+  if (r == 0)
```

Comment:
Minor: this is pretty repetitive for each step in the state machine.  Perhaps miove the logic to a common helper function so you can just run something like ```if (r == 0 && handle_cancellation()) { return; }```